tests/fe/common/fe_shape_function_plotter.h

```diff
@@ -0,0 +1,309 @@
+
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/numerics/vector_tools.h>
+
+#include <fstream>
+#include <iostream>
+#include <string>
+
+using namespace dealii;
+
+
+/**
+ * A helper struct that is used indicate whether we need to use
+ * FEExtractors in order to retrieve shape function information
+ * from the finite elements. This is typically required for finite
+ * elements with non-primitive shape functions.
+ */
+template <class FE_Type>
+struct RequiresFEExtractor : std::false_type
+{};
+
+/**
+ * A specialisation of the above classes, for the FE_NedelecSZ class.
+ */
+template <int dim, int spacedim>
+struct RequiresFEExtractor<FE_NedelecSZ<dim, spacedim>> : std::true_type
+{};
+
+
+/**
+ * A specialisation of the above classes, for the FE_RaviartThomas class.
+ */
+template <int dim>
+struct RequiresFEExtractor<FE_RaviartThomas<dim>> : std::true_type
+{};
+
+
+/**
+ * A function that returns the shape function of a finite element.
+ *
+ * This class uses directly requests the shape function components
+ * from the finite elements. Not all classes support this function
+ * call, so we have to take this into consideration.
+ */
+template <typename FE_Type>
+class FEComponentFunction : public Function<FE_Type::dimension>
+{
+public:
+  static const unsigned int dim = FE_Type::dimension;
+
+  FEComponentFunction(const FE_Type &      fe,
+                      const unsigned int   basis_function,
+                      const ReferenceCell &reference_cell)
+    : Function<dim>(fe.n_components())
+    , fe(fe)
+    , bf(basis_function)
+    , reference_cell(reference_cell)
+  {
+    Assert(basis_function < fe.n_dofs_per_cell(), ExcInternalError());
+  }
+  virtual ~FEComponentFunction() = default;
+
+  virtual void
+  vector_value(const Point<dim> &p, Vector<double> &values) const override
+  {
+    Assert(values.size() == this->n_components, ExcInternalError());
+
+    // Some functionality does not yet exist for transition elements
+    const bool is_pyramid = (reference_cell == ReferenceCells::Pyramid);
+    const bool is_wedge   = (reference_cell == ReferenceCells::Wedge);
+    if (!is_pyramid && !is_wedge)
+      {
+        Assert(reference_cell.contains_point(p), ExcInternalError());
+      }
+
+    // The main triangulation defines exactly the reference
+    // element, so we can therefore directly ask the FE
+    // for any shape functions etc.
+    for (unsigned int i = 0; i < this->n_components; ++i)
+      values[i] = fe.shape_value_component(bf, p, i);
+  }
+
+private:
+  const FE_Type &      fe;
+  const unsigned int   bf;
+  const ReferenceCell &reference_cell;
+};
+
+
+/**
+ * A function that returns the shape function of a finite element.
+ *
+ * This class uses FEExtractors to get the shape function information
+ * from the finite element. Doing so is significantly more expensive
+ * than the methodology used in the other variant of this class, so it
+ * should be used only when necessary.
```

Comment:
I think what you're trying to say is that it uses FEValues because the finite element does not implement the `shape_value()` function?